packages/experience-builder-sdk/src/hooks/useDetectMode.tsx

```diff
@@ -0,0 +1,103 @@
+  OUTGOING_EVENTS,
+  StudioExperienceMode,
+} from '@contentful/experiences-core/constants';
+import { sdkFeatures } from '../core/sdkFeatures';
+
+type UseDetectModeArgs = {
+  /** If running from a known client side only situation (ie: useFetchBySlug),
+   * set this to true to kick in editor mode check sooner (which avoids a render cycle) */
+  isClientSide?: boolean;
+};
+
+export const useDetectMode = ({ isClientSide = false }: UseDetectModeArgs = {}) => {
+  const [mounted, setMounted] = useState(false);
+  const receivedEditorModeMessage = useRef(false);
+  const receivedReadOnlyMode = useRef(false);
+  const [mode, setMode] = useState<StudioExperienceMode>(() => {
+    // if we are client side and running in an iframe, then initialize to read only,
+    // Editor mode can be requested later.
+    if (isClientSide && inIframe()) {
+      return StudioExperienceMode.READ_ONLY;
+    } else {
+      return StudioExperienceMode.NONE;
+    }
+  });
+
+  useEffect(() => {
+    const onMessage = (event: MessageEvent) => {
+      if (doesMismatchMessageSchema(event)) {
+        return;
+      }
+      const eventData = tryParseMessage(event);
+
+      if (eventData.eventType === INCOMING_EVENTS.RequestEditorMode) {
+        setMode(StudioExperienceMode.EDITOR);
+        receivedEditorModeMessage.current = true;
+        if (typeof window !== 'undefined') {
+          // Once we definitely know that we are in editor mode, we set this flag so future postMessage connect calls are not made
+          if (!window.__EB__) {
+            window.__EB__ = {};
+          }
+          window.__EB__.isReadOnlyMode = false;
+          window.__EB__.isEditorMode = true;
+        }
+      } else if (eventData.eventType === INCOMING_EVENTS.RequestReadOnlyMode) {
+        setMode(StudioExperienceMode.READ_ONLY);
+        receivedReadOnlyMode.current = true;
+        if (typeof window !== 'undefined') {
+          // Once we definitely know that we are in read only mode, we set this flag so future postMessage connect calls are not made
+          if (!window.__EB__) {
+            window.__EB__ = {};
+          }
+          window.__EB__.isEditorMode = false;
+          window.__EB__.isReadOnlyMode = true;
+        }
+      }
+
+      if (receivedEditorModeMessage.current && receivedReadOnlyMode.current) {
+        window.removeEventListener('message', onMessage);
+      }
+    };
+
+    // Only run check after component is mounted on the client to avoid hydration ssr issues
+    if (mounted) {
+      setMode(inIframe() ? StudioExperienceMode.EDITOR : StudioExperienceMode.NONE);
+      // Double check if we are in editor mode by listening to postMessage events
+      if (typeof window !== 'undefined') {
+        window.addEventListener('message', onMessage);
+        sendMessage(OUTGOING_EVENTS.Connected, undefined);
+        sendMessage(OUTGOING_EVENTS.SDKFeatures, sdkFeatures);
+
+        setTimeout(() => {
+          if (!receivedEditorModeMessage.current) {
```

Comment:
TODO: update this condition to see if either readonly, or editor mode have been requested.   If neither have been requested in 100MS, default to NONE.  this condition is currently checking only for editor mode.

- [ ] do it.